Mathlib/Tactic/Linter/DocString.lean

```diff
@@ -32,34 +51,42 @@ def docStringLinter : Linter where run := withSetOptionIn fun stx ↦ do
-  -- `docStx` extracts the `Lean.Parser.Command.docComment` node from the declaration modifiers.
-  -- In particular, this ignores parsing `#adaptation_note`s.
-  let docStx := declMods[0][0]
-  if docStx.isMissing then return
-  -- `docString` contains e.g. trailing spaces before the `-/`, but does not contain
-  -- any leading whitespace before the actual string starts.
-  let docString ← try getDocStringText ⟨docStx⟩ catch _ => return
-  -- `startSubstring` is the whitespace between `/--` and the actual doc-string text.
-  let startSubstring := match docStx with
-    | .node _ _ #[(.atom si ..), _] => si.getTrailing?.getD default
-    | _ => default
-  let start := startSubstring.toString
-  if !#["\n", " "].contains start then
-    let startRange := {start := startSubstring.startPos, stop := startSubstring.stopPos}
-    Linter.logLint linter.style.docString (.ofRange startRange)
-      s!"error: doc-strings should start with a single space or newline"
-  let docTrim := docString.trimRight
-  let tail := docTrim.length
-  -- `endRange` creates an 0-wide range `n` characters from the end of `docStx`
-  let endRange (n : Nat) : Syntax := .ofRange
-    {start := docStx.getTailPos?.get! - ⟨n⟩, stop := docStx.getTailPos?.get! - ⟨n⟩}
-  if docTrim.takeRight 1 == "," then
-    Linter.logLint linter.style.docString (endRange (docString.length - tail + 3))
-      s!"error: doc-strings should not end with a comma"
-  if tail + 1 != docString.length then
-    Linter.logLint linter.style.docString (endRange 3)
-      s!"error: doc-strings should end with a single space or newline"
+  let fm ← getFileMap
+  for declMods in getDeclModifiers stx do
+    -- `docStx` extracts the `Lean.Parser.Command.docComment` node from the declaration modifiers.
+    -- In particular, this ignores parsing `#adaptation_note`s.
+    let docStx := declMods[0][0]
+
+    let some pos := docStx.getPos? | continue
+    let startColumn := fm.toPosition pos |>.column
+
+    if docStx.isMissing then continue -- this is probably superfluous, thanks to `some pos` above.
+    -- `docString` contains e.g. trailing spaces before the `-/`, but does not contain
+    -- any leading whitespace before the actual string starts.
+    let docString ← try getDocStringText ⟨docStx⟩ catch _ => continue
+    -- `startSubstring` is the whitespace between `/--` and the actual doc-string text.
+    let startSubstring := match docStx with
+      | .node _ _ #[(.atom si ..), _] => si.getTrailing?.getD default
+      | _ => default
+    let start := deindentString startColumn startSubstring.toString
+    if !#["\n", " "].contains start then
+      let startRange := {start := startSubstring.startPos, stop := startSubstring.stopPos}
+      Linter.logLint linter.style.docString (.ofRange startRange)
+        s!"error: doc-strings should start with a single space or newline"
+
+    let deIndentedDocString := deindentString startColumn docString
```

Comment:
Inline this variable?

Reply:
I'd rather not: it is used twice, so mine is not a strong preference, though.